File: testdata/jennies/rawtypes/struct_with_map_and_slice_default/PHPRawTypes/src/Defaults/SomeStruct.php
<?php

namespace Grafana\Foundation\Defaults;

class SomeStruct implements \JsonSerializable
{
    /**
     * @var array<string, mixed>|null
     */
    public ?array $options;

    /**
     * @var array<string>|null
     */
    public ?array $items;

    /**
     * @var mixed
     */
    public $extra;

    /**
     * @param array<string, mixed>|null $options
     * @param array<string>|null $items
     * @param mixed|null $extra
     */
    public function __construct(?array $options = null, ?array $items = null,  $extra = null)
    {
        $this->options = $options;
        $this->items = $items;
        $this->extra = $extra ?: [];
    }

    public function equals(mixed $other): bool
    {
        if (!($other instanceof self)) {
            return false;
        }
    
        if (($this->options === null) !== ($other->options === null)) {
            return false;
        }
        if ($this->options !== null) {
            if ($this->options != $other->options) {
                return false;
            }
        }
    
        if (($this->items === null) !== ($other->items === null)) {
            return false;
        }
        if ($this->items !== null) {
            if ($this->items != $other->items) {
                return false;
            }
        }
    
        if ($this->extra !== $other->extra) {
            return false;
        }
    
        return true;
    }

    /**
     * @param array<string, mixed> $inputData
     */
    public static function fromArray(array $inputData): self
    {
        /** @var array{options?: array<string, mixed>, items?: array<string>, extra?: mixed} $inputData */
        $data = $inputData;
        return new self(
            options: $data["options"] ?? null,
            items: $data["items"] ?? null,
            extra: $data["extra"] ?? null,
        );
    }

    /**
     * @return mixed
     */
    public function jsonSerialize(): mixed
    {
        $data = new \stdClass;
        $data->extra = $this->extra;
        if (isset($this->options)) {
            $data->options = $this->options;
        }
        if (isset($this->items)) {
            $data->items = $this->items;
        }
        return $data;
    }
}

File: testdata/jennies/rawtypes/struct_with_map_and_slice_default/JavaRawTypes/defaults/SomeStruct.java
package defaults;

import java.util.Objects;
import java.util.Map;
import java.util.List;

public class SomeStruct {
    public Map<String, Object> options;
    public List<String> items;
    public Object extra;
    public SomeStruct() {
        this.options = Map.of();
        this.items = List.of();
        this.extra = Map.of();
    }
    public SomeStruct(Map<String, Object> options,List<String> items,Object extra) {
        this.options = options;
        this.items = items;
        this.extra = extra;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        if (!(other instanceof SomeStruct)) return false;
        SomeStruct o = (SomeStruct) other;
        if (!Objects.equals(this.options, o.options)) return false;
        if (!Objects.equals(this.items, o.items)) return false;
        if (!Objects.equals(this.extra, o.extra)) return false;
        return true;
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.options, this.items, this.extra);
    }
}
